// houndjs/lib/hound.js
'use strict';

var _MHSDK = require('./models/sdk/MHSDK.js');

var _MHSDK2 = _interopRequireDefault(_MHSDK);

var _MHObject = require('./models/base/MHObject.js');

var _MHObject2 = _interopRequireDefault(_MHObject);

var _MHAction = require('./models/action/MHAction.js');

var _MHAction2 = _interopRequireDefault(_MHAction);

var _MHAdd = require('./models/action/MHAdd.js');

var _MHAdd2 = _interopRequireDefault(_MHAdd);

var _MHComment = require('./models/action/MHComment.js');

var _MHComment2 = _interopRequireDefault(_MHComment);

var _MHCreate = require('./models/action/MHCreate.js');

var _MHCreate2 = _interopRequireDefault(_MHCreate);

var _MHLike = require('./models/action/MHLike.js');

var _MHLike2 = _interopRequireDefault(_MHLike);

var _MHFollow = require('./models/action/MHFollow.js');

var _MHFollow2 = _interopRequireDefault(_MHFollow);

var _MHPost = require('./models/action/MHPost.js');

var _MHPost2 = _interopRequireDefault(_MHPost);

var _MHHashtag = require('./models/hashtag/MHHashtag.js');

var _MHHashtag2 = _interopRequireDefault(_MHHashtag);

var _MHImage = require('./models/image/MHImage.js');

var _MHImage2 = _interopRequireDefault(_MHImage);

var _MHUser = require('./models/user/MHUser.js');

var _MHUser2 = _interopRequireDefault(_MHUser);

var _MHLoginSession = require('./models/user/MHLoginSession.js');

var _MHLoginSession2 = _interopRequireDefault(_MHLoginSession);

var _MHSocial = require('./models/social/MHSocial.js');

var _MHSocial2 = _interopRequireDefault(_MHSocial);

var _MHUserSocial = require('./models/social/MHUserSocial.js');

var _MHUserSocial2 = _interopRequireDefault(_MHUserSocial);

var _MHSocialMetrics = require('./models/social/MHSocialMetrics.js');

var _MHSocialMetrics2 = _interopRequireDefault(_MHSocialMetrics);

var _MHMedia = require('./models/media/MHMedia.js');

var _MHMedia2 = _interopRequireDefault(_MHMedia);

var _MHAlbum = require('./models/media/MHAlbum.js');

var _MHAlbum2 = _interopRequireDefault(_MHAlbum);

var _MHAlbumSeries = require('./models/media/MHAlbumSeries.js');

var _MHAlbumSeries2 = _interopRequireDefault(_MHAlbumSeries);

var _MHAnthology = require('./models/media/MHAnthology.js');

var _MHAnthology2 = _interopRequireDefault(_MHAnthology);

var _MHBook = require('./models/media/MHBook.js');

var _MHBook2 = _interopRequireDefault(_MHBook);

var _MHBookSeries = require('./models/media/MHBookSeries.js');

var _MHBookSeries2 = _interopRequireDefault(_MHBookSeries);

var _MHComicBook = require('./models/media/MHComicBook.js');

var _MHComicBook2 = _interopRequireDefault(_MHComicBook);

var _MHComicBookSeries = require('./models/media/MHComicBookSeries.js');

var _MHComicBookSeries2 = _interopRequireDefault(_MHComicBookSeries);

var _MHGame = require('./models/media/MHGame.js');

var _MHGame2 = _interopRequireDefault(_MHGame);

var _MHGameSeries = require('./models/media/MHGameSeries.js');

var _MHGameSeries2 = _interopRequireDefault(_MHGameSeries);

var _MHGraphicNovel = require('./models/media/MHGraphicNovel.js');

var _MHGraphicNovel2 = _interopRequireDefault(_MHGraphicNovel);

var _MHGraphicNovelSeries = require('./models/media/MHGraphicNovelSeries.js');

var _MHGraphicNovelSeries2 = _interopRequireDefault(_MHGraphicNovelSeries);

var _MHMovie = require('./models/media/MHMovie.js');

var _MHMovie2 = _interopRequireDefault(_MHMovie);

var _MHMovieSeries = require('./models/media/MHMovieSeries.js');

var _MHMovieSeries2 = _interopRequireDefault(_MHMovieSeries);

var _MHMusicVideo = require('./models/media/MHMusicVideo.js');

var _MHMusicVideo2 = _interopRequireDefault(_MHMusicVideo);

var _MHNovella = require('./models/media/MHNovella.js');

var _MHNovella2 = _interopRequireDefault(_MHNovella);

var _MHPeriodical = require('./models/media/MHPeriodical.js');

var _MHPeriodical2 = _interopRequireDefault(_MHPeriodical);

var _MHPeriodicalSeries = require('./models/media/MHPeriodicalSeries.js');

var _MHPeriodicalSeries2 = _interopRequireDefault(_MHPeriodicalSeries);

var _MHShowEpisode = require('./models/media/MHShowEpisode.js');

var _MHShowEpisode2 = _interopRequireDefault(_MHShowEpisode);

var _MHShowSeason = require('./models/media/MHShowSeason.js');

var _MHShowSeason2 = _interopRequireDefault(_MHShowSeason);

var _MHShowSeries = require('./models/media/MHShowSeries.js');

var _MHShowSeries2 = _interopRequireDefault(_MHShowSeries);

var _MHTrack = require('./models/media/MHTrack.js');

var _MHTrack2 = _interopRequireDefault(_MHTrack);

var _MHSpecial = require('./models/media/MHSpecial.js');

var _MHSpecial2 = _interopRequireDefault(_MHSpecial);

var _MHSpecialSeries = require('./models/media/MHSpecialSeries.js');

var _MHSpecialSeries2 = _interopRequireDefault(_MHSpecialSeries);

var _MHTrailer = require('./models/media/MHTrailer.js');

var _MHTrailer2 = _interopRequireDefault(_MHTrailer);

var _MHCollection = require('./models/collection/MHCollection.js');

var _MHCollection2 = _interopRequireDefault(_MHCollection);

var _MHSilo = require('./models/silo/MHSilo.js');

var _MHSilo2 = _interopRequireDefault(_MHSilo);

var _MHTrait = require('./models/trait/MHTrait.js');

var _MHTrait2 = _interopRequireDefault(_MHTrait);

var _MHGenre = require('./models/trait/MHGenre.js');

var _MHGenre2 = _interopRequireDefault(_MHGenre);

var _MHSubGenre = require('./models/trait/MHSubGenre.js');

var _MHSubGenre2 = _interopRequireDefault(_MHSubGenre);

var _MHMood = require('./models/trait/MHMood.js');

var _MHMood2 = _interopRequireDefault(_MHMood);

var _MHQuality = require('./models/trait/MHQuality.js');

var _MHQuality2 = _interopRequireDefault(_MHQuality);

var _MHStyleElement = require('./models/trait/MHStyleElement.js');

var _MHStyleElement2 = _interopRequireDefault(_MHStyleElement);

var _MHStoryElement = require('./models/trait/MHStoryElement.js');

var _MHStoryElement2 = _interopRequireDefault(_MHStoryElement);

var _MHMaterialSource = require('./models/trait/MHMaterialSource.js');

var _MHMaterialSource2 = _interopRequireDefault(_MHMaterialSource);

var _MHTheme = require('./models/trait/MHTheme.js');

var _MHTheme2 = _interopRequireDefault(_MHTheme);

var _MHAchievement = require('./models/trait/MHAchievement.js');

var _MHAchievement2 = _interopRequireDefault(_MHAchievement);

var _MHEra = require('./models/trait/MHEra.js');

var _MHEra2 = _interopRequireDefault(_MHEra);

var _MHAudience = require('./models/trait/MHAudience.js');

var _MHAudience2 = _interopRequireDefault(_MHAudience);

var _MHFlag = require('./models/trait/MHFlag.js');

var _MHFlag2 = _interopRequireDefault(_MHFlag);

var _MHGraphGenre = require('./models/trait/MHGraphGenre.js');

var _MHGraphGenre2 = _interopRequireDefault(_MHGraphGenre);

var _MHAudioTrait = require('./models/trait/MHAudioTrait.js');

var _MHAudioTrait2 = _interopRequireDefault(_MHAudioTrait);

var _MHLanguage = require('./models/trait/MHLanguage.js');

var _MHLanguage2 = _interopRequireDefault(_MHLanguage);

var _MHSubMood = require('./models/trait/MHSubMood.js');

var _MHSubMood2 = _interopRequireDefault(_MHSubMood);

var _MHSuitability = require('./models/trait/MHSuitability.js');

var _MHSuitability2 = _interopRequireDefault(_MHSuitability);

var _MHType = require('./models/trait/MHType.js');

var _MHType2 = _interopRequireDefault(_MHType);

var _MHContributor = require('./models/contributor/MHContributor.js');

var _MHContributor2 = _interopRequireDefault(_MHContributor);

var _MHRealIndividualContributor = require('./models/contributor/MHRealIndividualContributor.js');

var _MHRealIndividualContributor2 = _interopRequireDefault(_MHRealIndividualContributor);

var _MHRealGroupContributor = require('./models/contributor/MHRealGroupContributor.js');

var _MHRealGroupContributor2 = _interopRequireDefault(_MHRealGroupContributor);

var _MHFictionalIndividualContributor = require('./models/contributor/MHFictionalIndividualContributor.js');

var _MHFictionalIndividualContributor2 = _interopRequireDefault(_MHFictionalIndividualContributor);

var _MHFictionalGroupContributor = require('./models/contributor/MHFictionalGroupContributor.js');

var _MHFictionalGroupContributor2 = _interopRequireDefault(_MHFictionalGroupContributor);

var _MHSource = require('./models/source/MHSource.js');

var _MHSource2 = _interopRequireDefault(_MHSource);

var _MHSubscription = require('./models/source/MHSubscription.js');

var _MHSubscription2 = _interopRequireDefault(_MHSubscription);

var _MHSourceFormat = require('./models/source/MHSourceFormat.js');

var _MHSourceFormat2 = _interopRequireDefault(_MHSourceFormat);

var _MHSourceMethod = require('./models/source/MHSourceMethod.js');

var _MHSourceMethod2 = _interopRequireDefault(_MHSourceMethod);

var _MHSourceMedium = require('./models/source/MHSourceMedium.js');

var _MHSourceMedium2 = _interopRequireDefault(_MHSourceMedium);

var _MHSearch = require('./search/MHSearch.js');

var _MHSearch2 = _interopRequireDefault(_MHSearch);

var _houndRequest = require('./request/hound-request.js');

var _houndRequest2 = _interopRequireDefault(_houndRequest);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// We use a commonjs export here so ES5 clients don't have to do:
// require('houndjs').default
// Instead they can just do:
// require('houndjs')


// Search


// Contributor Models


// Media Models


// Action Models

// SDK
module.exports = {
  MHSDK: _MHSDK2.default,

  MHObject: _MHObject2.default,

  MHAction: _MHAction2.default,
  MHAdd: _MHAdd2.default,
  MHComment: _MHComment2.default,
  MHCreate: _MHCreate2.default,
  MHLike: _MHLike2.default,
  MHFollow: _MHFollow2.default,
  MHPost: _MHPost2.default,

  MHHashtag: _MHHashtag2.default,

  MHImage: _MHImage2.default,

  MHUser: _MHUser2.default,
  MHLoginSession: _MHLoginSession2.default,

  MHSocial: _MHSocial2.default,
  MHUserSocial: _MHUserSocial2.default,
  MHSocialMetrics: _MHSocialMetrics2.default,

  MHMedia: _MHMedia2.default,
  MHAlbum: _MHAlbum2.default,
  MHAlbumSeries: _MHAlbumSeries2.default,
  MHAnthology: _MHAnthology2.default,
  MHBook: _MHBook2.default,
  MHBookSeries: _MHBookSeries2.default,
  MHComicBook: _MHComicBook2.default,
  MHComicBookSeries: _MHComicBookSeries2.default,
  MHGame: _MHGame2.default,
  MHGameSeries: _MHGameSeries2.default,
  MHGraphicNovel: _MHGraphicNovel2.default,
  MHGraphicNovelSeries: _MHGraphicNovelSeries2.default,
  MHMovie: _MHMovie2.default,
  MHMovieSeries: _MHMovieSeries2.default,
  MHMusicVideo: _MHMusicVideo2.default,
  MHNovella: _MHNovella2.default,
  MHPeriodical: _MHPeriodical2.default,
  MHPeriodicalSeries: _MHPeriodicalSeries2.default,
  MHShowEpisode: _MHShowEpisode2.default,
  MHShowSeason: _MHShowSeason2.default,
  MHShowSeries: _MHShowSeries2.default,
  MHTrack: _MHTrack2.default,
  MHSpecial: _MHSpecial2.default,
  MHSpecialSeries: _MHSpecialSeries2.default,
  MHTrailer: _MHTrailer2.default,

  MHCollection: _MHCollection2.default,

  MHSilo: _MHSilo2.default,

  MHTrait: _MHTrait2.default,
  MHGenre: _MHGenre2.default,
  MHSubGenre: _MHSubGenre2.default,
  MHMood: _MHMood2.default,
  MHQuality: _MHQuality2.default,
  MHStyleElement: _MHStyleElement2.default,
  MHStoryElement: _MHStoryElement2.default,
  MHMaterialSource: _MHMaterialSource2.default,
  MHTheme: _MHTheme2.default,
  MHAchievement: _MHAchievement2.default,
  MHEra: _MHEra2.default,
  MHAudience: _MHAudience2.default,
  MHFlag: _MHFlag2.default,
  MHGraphGenre: _MHGraphGenre2.default,

  MHContributor: _MHContributor2.default,
  MHRealIndividualContributor: _MHRealIndividualContributor2.default,
  MHRealGroupContributor: _MHRealGroupContributor2.default,
  MHFictionalIndividualContributor: _MHFictionalIndividualContributor2.default,
  MHFictionalGroupContributor: _MHFictionalGroupContributor2.default,

  MHSource: _MHSource2.default,
  MHSubscription: _MHSubscription2.default,
  MHSourceFormat: _MHSourceFormat2.default,
  MHSourceMethod: _MHSourceMethod2.default,
  MHSourceMedium: _MHSourceMedium2.default,

  MHSearch: _MHSearch2.default,

  houndRequest: _houndRequest2.default
};

// Request


// Source Models


// Trait Models


// Social Models


// User Models


// Hashtag Models


// Base Models
