// houndjs/lib/models/meta/MHMetadata.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MHTraitMetadata = exports.MHHashtagMetadata = exports.MHContributorMetadata = exports.MHSourceMetadata = exports.MHSubscriptionMetadata = exports.MHImageMetadata = exports.MHActionMetadata = exports.MHCollectionMetadata = exports.MHUserMetadata = exports.MHMediaMetadata = exports.MHMetadata = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _MHImageData = require('../image/MHImageData.js');

var _MHImageData2 = _interopRequireDefault(_MHImageData);

var _jsonParse = require('../internal/jsonParse.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var MHMetadata = exports.MHMetadata = function () {
  function MHMetadata(args) {
    _classCallCheck(this, MHMetadata);

    (0, _jsonParse.jsonCreateWithArgs)(args, this);
  }

  _createClass(MHMetadata, [{
    key: 'jsonProperties',
    get: function get() {
      return {
        mhid: { type: String, optional: false },
        altId: String,
        name: String,
        description: String,
        createdDate: Date,
        subType: String
      };
    }
  }]);

  return MHMetadata;
}();

var MHMediaMetadata = exports.MHMediaMetadata = function (_MHMetadata) {
  _inherits(MHMediaMetadata, _MHMetadata);

  function MHMediaMetadata() {
    _classCallCheck(this, MHMediaMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHMediaMetadata).apply(this, arguments));
  }

  _createClass(MHMediaMetadata, [{
    key: 'jsonProperties',
    get: function get() {
      return _extends({}, _get(Object.getPrototypeOf(MHMediaMetadata.prototype), 'jsonProperties', this), {
        releaseDate: Date
      });
    }
  }]);

  return MHMediaMetadata;
}(MHMetadata);

var MHUserMetadata = exports.MHUserMetadata = function (_MHMetadata2) {
  _inherits(MHUserMetadata, _MHMetadata2);

  function MHUserMetadata() {
    _classCallCheck(this, MHUserMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHUserMetadata).apply(this, arguments));
  }

  _createClass(MHUserMetadata, [{
    key: 'jsonProperties',
    get: function get() {
      return _extends({}, _get(Object.getPrototypeOf(MHUserMetadata.prototype), 'jsonProperties', this), {
        username: { type: String, optional: false },
        email: String,
        firstName: String,
        lastName: String
      });
    }
  }]);

  return MHUserMetadata;
}(MHMetadata);

var MHCollectionMetadata = exports.MHCollectionMetadata = function (_MHMetadata3) {
  _inherits(MHCollectionMetadata, _MHMetadata3);

  function MHCollectionMetadata() {
    _classCallCheck(this, MHCollectionMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHCollectionMetadata).apply(this, arguments));
  }

  _createClass(MHCollectionMetadata, [{
    key: 'jsonProperties',
    get: function get() {
      return _extends({}, _get(Object.getPrototypeOf(MHCollectionMetadata.prototype), 'jsonProperties', this), {
        mixlist: String
      });
    }
  }]);

  return MHCollectionMetadata;
}(MHMetadata);

var MHActionMetadata = exports.MHActionMetadata = function (_MHMetadata4) {
  _inherits(MHActionMetadata, _MHMetadata4);

  function MHActionMetadata() {
    _classCallCheck(this, MHActionMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHActionMetadata).apply(this, arguments));
  }

  _createClass(MHActionMetadata, [{
    key: 'jsonProperties',
    get: function get() {
      return _extends({}, _get(Object.getPrototypeOf(MHActionMetadata.prototype), 'jsonProperties', this), {
        message: String
      });
    }
  }]);

  return MHActionMetadata;
}(MHMetadata);

var MHImageMetadata = exports.MHImageMetadata = function (_MHMetadata5) {
  _inherits(MHImageMetadata, _MHMetadata5);

  function MHImageMetadata() {
    _classCallCheck(this, MHImageMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHImageMetadata).apply(this, arguments));
  }

  _createClass(MHImageMetadata, [{
    key: 'jsonProperties',
    get: function get() {
      return _extends({}, _get(Object.getPrototypeOf(MHImageMetadata.prototype), 'jsonProperties', this), {
        isDefault: Boolean,
        averageColor: String,
        thumbnail: _MHImageData2.default,
        small: _MHImageData2.default,
        medium: _MHImageData2.default,
        large: _MHImageData2.default,
        original: _MHImageData2.default
      });
    }
  }]);

  return MHImageMetadata;
}(MHMetadata);

var MHSubscriptionMetadata = exports.MHSubscriptionMetadata = function (_MHMetadata6) {
  _inherits(MHSubscriptionMetadata, _MHMetadata6);

  function MHSubscriptionMetadata() {
    _classCallCheck(this, MHSubscriptionMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHSubscriptionMetadata).apply(this, arguments));
  }

  _createClass(MHSubscriptionMetadata, [{
    key: 'jsonProperties',
    get: function get() {
      return _extends({}, _get(Object.getPrototypeOf(MHSubscriptionMetadata.prototype), 'jsonProperties', this), {
        timePeriod: String,
        price: String,
        currency: String,
        mediums: String
      });
    }
  }]);

  return MHSubscriptionMetadata;
}(MHMetadata);

var MHSourceMetadata = exports.MHSourceMetadata = function (_MHMetadata7) {
  _inherits(MHSourceMetadata, _MHMetadata7);

  function MHSourceMetadata() {
    _classCallCheck(this, MHSourceMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHSourceMetadata).apply(this, arguments));
  }

  _createClass(MHSourceMetadata, [{
    key: 'jsonProperties',
    get: function get() {
      return _extends({}, _get(Object.getPrototypeOf(MHSourceMetadata.prototype), 'jsonProperties', this), {
        connectable: Boolean
      });
    }
  }]);

  return MHSourceMetadata;
}(MHMetadata);

var MHContributorMetadata = exports.MHContributorMetadata = function (_MHMetadata8) {
  _inherits(MHContributorMetadata, _MHMetadata8);

  function MHContributorMetadata() {
    _classCallCheck(this, MHContributorMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHContributorMetadata).apply(this, arguments));
  }

  return MHContributorMetadata;
}(MHMetadata);

var MHHashtagMetadata = exports.MHHashtagMetadata = function (_MHMetadata9) {
  _inherits(MHHashtagMetadata, _MHMetadata9);

  function MHHashtagMetadata() {
    _classCallCheck(this, MHHashtagMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHHashtagMetadata).apply(this, arguments));
  }

  return MHHashtagMetadata;
}(MHMetadata);

var MHTraitMetadata = exports.MHTraitMetadata = function (_MHMetadata10) {
  _inherits(MHTraitMetadata, _MHMetadata10);

  function MHTraitMetadata() {
    _classCallCheck(this, MHTraitMetadata);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHTraitMetadata).apply(this, arguments));
  }

  _createClass(MHTraitMetadata, [{
    key: 'jsonProperties',
    get: function get() {
      return _extends({}, _get(Object.getPrototypeOf(MHTraitMetadata.prototype), 'jsonProperties', this), {
        pluralName: String
      });
    }
  }]);

  return MHTraitMetadata;
}(MHMetadata);
