// houndjs/lib/models/hashtag/MHHashtag.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _MHObject2 = require('../base/MHObject.js');

var _MHObject3 = _interopRequireDefault(_MHObject2);

var _houndRequest = require('../../request/hound-request.js');

var _houndRequest2 = _interopRequireDefault(_houndRequest);

var _MHMetadata = require('../meta/MHMetadata.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MHHashtag = function (_MHObject) {
  _inherits(MHHashtag, _MHObject);

  function MHHashtag() {
    _classCallCheck(this, MHHashtag);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(MHHashtag).apply(this, arguments));
  }

  _createClass(MHHashtag, [{
    key: 'jsonProperties',
    get: function get() {
      return _extends({}, _get(Object.getPrototypeOf(MHHashtag.prototype), 'jsonProperties', this), {
        metadata: _MHMetadata.MHHashtagMetadata
      });
    }

    /**
    * MHHashtag.fetchByName(name,view,force)
    *
    * @param { String } username - Username to fetch info for
    * @param { boolean} force - force fetch to server
    *
    * @return { Promise } - resolves to the MHUser object
    *
    */

  }], [{
    key: 'fetchByName',
    value: function fetchByName(name) /*, force=false*/{
      var view = arguments.length <= 1 || arguments[1] === undefined ? 'full' : arguments[1];

      if (!name || typeof name !== 'string' && !(name instanceof String)) {
        throw new TypeError('Hashtag not of type String in fetchByTag');
      }

      var path = this.rootSubendpoint('/lookup/' + name);

      return (0, _houndRequest2.default)({
        method: 'GET',
        endpoint: path,
        params: {
          view: view
        }
      }).then(function (response) {
        var newObj = _MHObject3.default.create(response);
        return newObj;
      });
    }
  }, {
    key: 'mhidPrefix',
    get: function get() {
      return 'mhhtg';
    }
  }, {
    key: 'rootEndpoint',
    get: function get() {
      return 'graph/hashtag';
    }
  }]);

  return MHHashtag;
}(_MHObject3.default);

exports.default = MHHashtag;


_MHObject3.default.registerConstructor(MHHashtag, 'MHHashtag');
