// houndjs/lib/models/container/MHContext.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _jsonParse = require('../internal/jsonParse.js');

var _MHRelationship = require('./MHRelationship.js');

var _MHRelationship2 = _interopRequireDefault(_MHRelationship);

var _MHRelationalPair = require('./MHRelationalPair.js');

var _MHRelationalPair2 = _interopRequireDefault(_MHRelationalPair);

var _MHSorting = require('./MHSorting.js');

var _MHSorting2 = _interopRequireDefault(_MHSorting);

var _MHSourceMedium = require('../source/MHSourceMedium.js');

var _MHSourceMedium2 = _interopRequireDefault(_MHSourceMedium);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var MHContext = function () {
  function MHContext(args) {
    _classCallCheck(this, MHContext);

    (0, _jsonParse.jsonCreateWithArgs)(args, this);
  }

  _createClass(MHContext, [{
    key: 'jsonProperties',
    get: function get() {
      return {
        metadata: Object,
        consumable: Boolean,
        sorting: _MHSorting2.default,
        fixed: _MHSorting2.default,
        relationships: [_MHRelationship2.default],
        mediums: [_MHSourceMedium2.default],
        difference: [_MHRelationalPair2.default],
        social: Object,
        launchInfo: Object
      };
    }
  }]);

  return MHContext;
}();

exports.default = MHContext;
